### PR TITLE
added safe option to s3store to prevent unintentional overwrites

### PR DESCRIPTION
**Pull request by kirillseva**

### CODE REVIEW

**Comment by rylandely** [comment]
We've needed this for a long time.  Thanks for adding it.

**Comment by iveksl2** [comment]
Might want to then add a flag to export stage. If one runs a full syberia file and it fails to export because your overwriting an object, the whole syberia-run will need to be rerun.

**Comment by peterhurford** [comment]
This is a good implementation, but I think it's going to break a lot of things in other packages as this is not backwards compatible.  We'll have to think what we want to do about that.

**Comment by kirillseva** [comment]
@peterhurford which packages? There isn't a HUGE ecosystem yet, so we can rewrite where needed.
most of the time it will break things that should be broken.
Or we can change the safe option to be opt-in instead of opt-out, but that would be less useful.
@iveksl2 I'll submit a PR for the s3 adapter to support the safe option as well.

**Comment by kirillseva** [comment]
bumped the package version so that we can enforce it later through lockfile.
The new syntax for syberia export stage will be
```r
export = list(
    s3=list(name="syberia/titanic/glm", safe = TRUE) # will not overwrite, opt-in
    # s3="syberia/titanic/glm" # will overwrite, default behavior is unchanged
  )
```

File: DESCRIPTION [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Package: s3mpi
 Title: R message passing interface using S3 storage
 Description: Easily pass objects like lists or dataframes between consoles
-Version: 0.2.1
+Version: 0.3.0
```
Comment (robertzk) [inline DESCRIPTION]: 0.2.2 please

File: R/s3store.r [hunk 2]
```diff
@@ -9,10 +9,15 @@
 #' s3store(c(1,2,3), 'test123')
 #' print(s3read('test123'))
 #' # [1] 1 2 3
-#' }#' 
-s3store <- function(obj, name = NULL, .path = s3path(), ...) {
+#' }#'
+s3store <- function(obj, name = NULL, .path = s3path(), safe = TRUE, ...) {
   if (is.null(name)) name <- deparse(substitute(obj))
   s3key <- paste(.path, name, sep = '')
+  if (safe && s3mpi::s3exists(name, .path = s3path(), ...)) {
```
Comment (robertzk) [inline R/s3store.r]: You need `.path = .path`

File: R/s3store.r [hunk 3]
```diff
@@ -9,10 +9,15 @@
...
+    # using cat prints to stdout as opposed to messages, so it can be seen from syberia::run_model()
+    cat("An object with this name already exists. Use `safe = FALSE` to overwrite\n")
+    stop("------------------^")
```
Comment (robertzk) [inline R/s3store.r]: Just put this in the stop message.

Comment (robertzk) [inline R/s3store.r]: Also write out the actual name of the key in the message -- what if this is from deep in a call stack? The user won't know what is happening.

**Comment by robertzk** [comment]
Please update the `NEWS.md` with documentation on the new option. Use 0.2.2 instead of 0.3.0 since the change is not major.

File: R/s3store.r (hunk 3 above)
Comment (kirillseva) [inline R/s3store.r]: The error message inside the ```stop()``` is not shown when s3store is used in export stage of a syberia run. ```cat``` is shown though

**Comment by robertzk** [comment]
Nice!

**Pull request merged**